File: daily_python/3532_Path_Existence_Queries_in_a_Graph_I.py
import os
import sys
from typing import List

class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[bool]:
        farrest = []
        curr = 0
        for i in range(n):
            curr = max(curr, i)
            while curr < n - 1 and nums[curr + 1] - nums[curr] <= maxDiff:
                curr += 1
            farrest.append(curr)
        ans = []
        for query in queries:
            s = min(query[0], query[1])
            e = max(query[0], query[1])
            ans.append(farrest[s] >= e)
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            n = int(lines[0].split('=')[1].strip()[:-1])
            nums = list(map(int, lines[1].split('=')[1].strip()[1:-2].split(',')))
            maxDiff = int(lines[2].split('=')[1].strip()[:-1])
            queries = [list(map(int, building.split(','))) for building in lines[3].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append((n, nums, maxDiff, queries))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for n, nums, maxDiff, queries in dataset:
        results.append(solution.pathExistenceQueries(n, nums, maxDiff, queries))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2026/3532_Path_Existence_Queries_in_a_Graph_I.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<bool> pathExistenceQueries(int n, vector<int>& nums, int maxDiff, vector<vector<int>>& queries) {
        vector<int> farrest(n, 0);
        int curr = 0;
        for (int i = 0; i < n; i++){
            curr = max(curr, i);
            while(curr < n - 1 && nums[curr + 1] - nums[curr] <= maxDiff) curr++;
            farrest[i] = curr;
        }
        int q = queries.size();
        vector<bool> ans(q);
        for (int i = 0; i < q; i++){
            int s = min(queries[i][0], queries[i][1]);
            int e = max(queries[i][0], queries[i][1]);
            ans[i] = (farrest[s] >= e);
        }
        return ans;
    }
};

int main(){
    int n = 2;
    vector<int> nums = {1,3};
    int maxDiff = 1;
    vector<vector<int>> queries = {{0,0},{0,1}};
    // int n = 4;
    // vector<int> nums = {2,5,6,8};
    // int maxDiff = 2;
    // vector<vector<int>> queries = {{0,1},{0,2},{1,3},{2,3}};
    // int n = 2;
    // vector<int> nums = {2975,50642};
    // int maxDiff = 6;
    // vector<vector<int>> queries = {{1,0}};
    Solution S;
    vector<bool> ans = S.pathExistenceQueries(n, nums, maxDiff, queries);
    for (bool i : ans) cout << i << " ";
    cout << endl;
    return 0;
}